libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -1790,6 +1817,19 @@ simplifyPred = \case
     x :-> p | Just t <- pinnedBy x p -> simplifyPred $ substitutePred x t p
     b' -> Exists k b'
   Monitor {} -> TruePred
+  -- TODO: This is a bit questionable. On the one hand getting rid of `Explain` does
+  -- simplfy the constraints to the point where we can use them, on the other hand
+  -- not getting rid of `explain` breaks a bunch of optimizations in bad ways. I don't
+  -- know what to do.
+  Explain es p ->
+    let p' = simplifyPred p
+     in case simplifyPred p of
+          FalsePred es' -> FalsePred (es ++ es')
+          TruePred {} -> p'
+          GenHint {} -> p'
+          Monitor {} -> p'
+          Explain es' p'' -> Explain (es ++ es') p''
+          p' -> Explain es p'
```

Comment:
It looks like, you are not getting rid of Explain, just simplifying underneath. What would getting rid of it look like? I am not sure I understand the breadth of this comment.

Reply:
I've updated the comment to be more clear. The issue is that it would be nice to just drop `Explain es` here and return `simplifyPred p`, but that risks creating an issue where explanations are removed too soon.